staging/src/k8s.io/component-base/config/serializer/decode.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package serializer
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/apimachinery/pkg/runtime/serializer/recognizer"
+)
+
+// strictSerializerExtensions are for serializers that are conditionally compiled in.
+// Note that this is a separate variable from serializer.CodecFactory.
+var strictSerializerExtensions = []func(*runtime.Scheme) (serializerType, bool){}
+
+// NewStrictUniversalDecoder returns a runtime.Decoder which operates in a strict manner.
+// This Decoder is capable of decoding all known API objects in all known formats. It is used by clients that do not need
+// to encode objects but want to deserialize API objects stored on disk. It only decodes objects in groups registered
+// with the scheme. GroupVersions may be passed in `versions` to specify alternate versions of objects to return.
+// Populating `versions` is optional; a zero-length `versions` will result in runtime.InternalGroupVersioner -- this matches
+// behavior of serializer.CodecFactory.UniversalDecoder().
+// If any versions are specified, unrecognized groups will be returned in the version they are encoded as (no conversion).
+// This decoder performs defaulting.
+//
+// This constructor is not included in serializer.CodecFactory because all runtime Decode operations are intended to be strict in the future.
+// The current strict checking implementation is not fit for the hot-path of the apiserver due to performance reasons.
+// This constructor is a temporary solution until we can rely on serializer.CodecFactory.UniversalDecoder() to return Strict errors.
+//
+// TODO: the decoder will eventually be removed in favor of dealing with objects in their versioned form
+// TODO: only accept a group versioner
+func NewStrictUniversalDecoder(scheme *runtime.Scheme, versions []schema.GroupVersion) runtime.Decoder {
```

Comment:
I would make the existing NewUniversalDecoder take and pass through the options struct rather than copying this.

(In practice you may have to e.g. make a NewUniversalDecoderWithOptions so as to reduce the code change, but you can modify the existing function to call the new function, so there's no code duplication and test coverage is at least as good as it was.)

Reply:
I like this suggestion a lot -- thanks.